### ISSUE TITLE
Consider upgrading MongoDB to 3.2

### ISSUE THREAD

**Issue by avital** [meteor/meteor#5809]
Please see https://github.com/meteor/meteor/issues/6957 for a high level description of what needs to be done on this ticket. 

---

Time estimate: 1 to 4 days.

Looks like we need to upgrade the driver, and parts of our `mongo` code. And upgrade to Node 4.

See https://github.com/meteor/meteor/issues/5763

**Comment by zimme** [comment]
As long as the driver, the `mongo` package and node is updated we can use meteor with the latest mongodb. You could still stay on any previous version of mongodb for meteor though, as the new driver still works with older versions of mongodb.

**Comment by avital** [comment]
We're not going to upgrade to Node 4 in Meteor 1.3 -- there is already a lot of valuable work in 1.3 and we should get it out soon so people can start using its new features. We should consider upgrading Node in the next release afterwards.

The main challenge to upgrading Node is supporting a new v8 embedding API, meaning that all old packages with binary dependencies would need to be re-published.

**Comment by serkandurusoy** [comment]
@avital I believe there's also a problem with fibers' binary availability which may need to be addressed first.

**Comment by avital** [comment]
Hmm. Maybe we don't need to upgrade to Node 4 to upgrade the Mongo driver? https://github.com/meteor/meteor/issues/5763#issuecomment-164947624

**Comment by NicolasParada** [comment]
Yep, https://docs.mongodb.org/ecosystem/drivers/node-js/ mongo 3.2 need driver version >= 2.1 and these driver is supported even in node 0.8

**Comment by bySabi** [comment]
That´s correct !! ... is my fault cause I don´t parse driver Language Compatibility table correctly.
Driver 2.1 is supported on Node 0.10

@avital What packages have binary dependencies?

**Comment by zimme** [comment]
@serkandurusoy, is that really a problem? I thought @benjamn was getting rid of `node-fibers` in favour of promises.

**Comment by serkandurusoy** [comment]
Wow I missed that news and if that's the case it of happily course is not!

Sent from my iPhone



On Dec 16, 2015, at 11:39, Simon Fridlund <notifications@github.com> wrote:

@serkandurusoy <https://github.com/serkandurusoy>, is that really a
problem? I thought @benjamn <https://github.com/benjamn> was getting rid of
node-fibers in favour of promises.

—
Reply to this email directly or view it on GitHub
<https://github.com/meteor/meteor/issues/5809#issuecomment-165048081>.

**Comment by zimme** [comment]
I've seem a lot of these commits: Eliminate Future from tools/files.js. from benjamn

**Comment by serkandurusoy** [comment]
Yeah I vaguely remember that from when he first took a stab at these but I thought that was not prioritized. Nice to know.

**Comment by avital** [comment]
The changes @benjamn made eliminated Futures in favor of Promises that have a `await` method. The `await` method is still implemented with Fibers. A future version of Meteor may no longer use Fibers but we're not there yet.

Luckily, looks like [fibers aren't an issue with Node 4](https://github.com/meteor/meteor/issues/5124#issuecomment-141677580) though.

We'll try upgrading to Mongo 3.2 for Meteor 1.3.

Let's keep this thread on-topic.

**Comment by serkandurusoy** [comment]
Is this the right spot to discuss an api to better integrate with mongodb's
* schema validation feature
* partial indexes

both of which landed with 3.2? Or would you rather see a community package that relies on rawCollection to handle those?

**Comment by avital** [comment]
Can you open a separate issue to discuss those?

It's plausible that Meteor 1.3 will ship with support for the new mongo driver without supplying a Meteor API for those features, since as you point out it's easy to do with rawCollection.

**Comment by zimme** [comment]
I would love to see the newer driver in 1.3, even if it's without an api for the new features; as you have access to them through `rawCollection`/`rawDatabase`.

**Comment by fabiodr** [comment]
Great! Just getting the newest driver is a big advance, i suspect we get better BSON serialization performance, which can make a big difference for large data sets :+1:

**Comment by zimme** [comment]
The only mention of the newer driver being needed for MongoDB 3.2 in the release notes are
> readConcern requires MongoDB drivers updated for MongoDB 3.2

So maybe a newer driver isn't really needed to upgrade to 3.2?

**Comment by serkandurusoy** [comment]
@zimme but then what do you make of the [compatibility chart](https://docs.mongodb.org/ecosystem/drivers/node-js/#mongodb-compatibility)?

**Comment by zimme** [comment]
well, >= 2.0.14 isn't fully compatible with mongo 3.2 because of not supporting `readConcern`.

I'll try and play around with it some more later today and see, because I think I was having problems with mongo 3.2.0 before and I think it works with mongo 3.2.1, but I'm not sure.

**Comment by mcchrish** [comment]
Compose.io recently updated their MongoDB service to 3.2.1 ([link](https://www.compose.io/articles/composes-new-mongodb-and-what-you-need-to-know/)). Now I'm looking into making it work with my current Meteor project, but I'm getting the following error related to the mongo driver:
```
W20160209-17:24:46.086(8)? (STDERR) /Users/mclopez/.meteor/packages/meteor-tool/.1.1.10.nlhe7j++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:278
W20160209-17:24:46.086(8)? (STDERR)                                             throw(ex);
W20160209-17:24:46.086(8)? (STDERR)                                                   ^
W20160209-17:24:46.195(8)? (STDERR) TypeError: Cannot read property 'master' of undefined
W20160209-17:24:46.196(8)? (STDERR)     at Object.Future.wait (/Users/mclopez/.meteor/packages/meteor-tool/.1.1.10.nlhe7j++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:398:15)
W20160209-17:24:46.196(8)? (STDERR)     at new MongoConnection (packages/mongo/mongo_driver.js:213:1)
W20160209-17:24:46.196(8)? (STDERR)     at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:1)
W20160209-17:24:46.196(8)? (STDERR)     at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:1)
W20160209-17:24:46.196(8)? (STDERR)     at Object.defaultRemoteCollectionDriver (packages/underscore/underscore.js:750:1)
W20160209-17:24:46.196(8)? (STDERR)     at new Mongo.Collection (packages/mongo/collection.js:102:1)
W20160209-17:24:46.196(8)? (STDERR)     at AccountsServer.AccountsCommon (accounts_common.js:23:18)
W20160209-17:24:46.196(8)? (STDERR)     at new AccountsServer (accounts_server.js:16:5)
W20160209-17:24:46.196(8)? (STDERR)     at Package (globals_server.js:5:12)
W20160209-17:24:46.196(8)? (STDERR)     at /Users/mclopez/inspirit-project/inspirit/.meteor/local/build/programs/server/packages/accounts-base.js:1814:4
W20160209-17:24:46.196(8)? (STDERR)     - - - - -
W20160209-17:24:46.196(8)? (STDERR)     at packages/mongo/mongo_driver.js:183:1
W20160209-17:24:46.196(8)? (STDERR)     at runWithEnvironment (packages/meteor/dynamics_nodejs.js:110:1)
=> Exited with code: 8
```
I'm currently not sure if this is cause by my current setup. Also compose.io's architecture for Mongo has considerably change that might be also the cause.

For now compose.io still offers their "MongoDB Classic" which is their older type of service and also with a lower version of mongo. I might try to spin up a classic one and try if it works.

**Comment by ljack** [comment]
Hello,
 There's also the slight issue of meteor <= 1.2 with older mongo (2.x?)  being unable to connect .meteor.com hosted apps. So it seems meteor.com hosting is already using mongo 3.x ?

**Comment by zol** [comment]
We should ensure we support SSL cert validation: http://mongodb.github.io/node-mongodb-native/2.0/tutorials/enterprise_features/?_ga=1.126504205.1665209564.1436572788

**Comment by oceddi** [comment]
I agree with zol.  You should take this opportunity to support SSL client authentication (not just server authentication) by letting users specify the "server" parameter to the MongoDB connect call.  That way the client key and certificate (and CA cert) can be specified and we can turn on client auth enforcement on the Mongo side.

**Comment by dgtlife** [comment]
I have my replica set humming along with X.509 on SSL, and local Mongo Shell Admin users logging in with X.509 user strings ... only to be stuck now with Meteor 1.3.x apparently not capable of supporting this. So, I would also like to see X.509 Auth + SSL for the Meteor app. Can anyone hazard a guess as to when this might materialize? I may have to hack on the driver to get some joy.

**Comment by abernix** [comment]
@dgtlife As soon as possible!  Help with the driver if you can!   Details in #6957.

**Comment by dgtlife** [comment]
@abernix I guess "short/medium" term is all we have as a forecast right now. I will prioritize my app features above this driver though, and if I get to the point where I have to hack it, then it will be just that ... a hack to bridge the gap until MDG delivers.

**Comment by bradvogel** [comment]
Just a note: if you're not using oplog tailing and are OK with forking the accounts package (which does its own db calls), you can just pull in the newer driver via NPM and use it directly

**Comment by dgtlife** [comment]
@bradvogel Have you done this ... or just imagining? ;-)

**Comment by bradvogel** [comment]
In the process of doing it - works so far on Galaxy!

**Comment by dgtlife** [comment]
I see. Well, that's encouraging. Thanks for the pointer.

**Comment by zol** [comment]
@bradvogel would love to see some code when you're finished.

**Comment by bradvogel** [comment]
Note for all considering upgrading to Mongo 3.2: bug https://jira.mongodb.org/browse/SERVER-14322 is a pretty big gotcha. If you're upserting (or its findAndModify equivalent) a document, that document might get duplicated (if the index in non-unique) or throw an exception (if the index is unique). We're seeing quite a few of these.

**Comment by dgtlife** [comment]
OK, I've got this working too:
MongoDB v3.2.6,
MongoDB Node.js Driver v2.1.18,
X.509 Auth on SSL for default connection,
X.509 Auth on SSL for Oplog Tailing connection (verified with the facts package).

Thanks again for the pointer @bradvogel

**Comment by abernix** [comment]
@dgtlife Glad to hear you got it working.  Please consider commenting on #6957 with any issues you ran into and any recommendations you might have, to help get the rest of the Meteor community there.  (Yes, that issue has the `pull-requests-encouraged` label!)

**Comment by dgtlife** [comment]
Given the larger and more comprehensive scope described in #6957, I'm not sure my changes would fit nicely into the eventual approach. For example, I chose to retain `npm-mongo` for the sake of minimal disturbance to Meteor 1.3.x, and Tom et al. are considering life without it. It may be better/more efficient if I just post my modified packages and allow the code to be cherry picked from there as desired. I'll post a link within the next 24 hours.

**Comment by dgtlife** [comment]
Here is the link to the modified `mongo` and `npm-mongo` that work for me [https://github.com/dgtlife/meteor-mongo-for-3.2.x](url)

**Comment by Hongbo-Miao** [comment]
@dgtlife just remind your link to a wrong link

**Comment by dgtlife** [comment]
@Hongbo-Miao Sorry, it's good now.

**Comment by jedwards1211** [comment]
MongoDB 3.2 has been working with some of our Meteor projects (some using Meteor 1.0.5, some using 1.2.1) that were built with `demeteorizer`, but it's not working with another project and I am confused as f*** why

**Comment by jedwards1211** [comment]
Appears it worked with 3.2.4 without any special intervention, but not with 3.2.0.

**Comment by interfaith** [comment]
is there anyway to run the latest meteor  with mongo 3.2.7 ?

**Comment by mhagmajer** [comment]
+1

**Comment by voidale** [comment]
+1

**Comment by laosb** [comment]
Fixed in 1.4. Closing.
